File: C++/1105. Filling Bookcase Shelves.cpp
class Solution {
public:
    int dp[1001][1001] = {};
    int solve(vector<vector<int>>& books, int shelfWidth, int index, int width, int height){
        if(index >= books.size()){
            return height;
        }
        if(dp[index][width] == 0){
            dp[index][width] = height+ solve(books, shelfWidth, index + 1, books[index][0], books[index][1]);
            if(width + books[index][0] <= shelfWidth){
                dp[index][width] = min(dp[index][width], solve(books, shelfWidth, index + 1, width + books[index][0], max(height, books[index][1])));
            }
        }
        return dp[index][width];
    }
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        int ans = solve(books, shelfWidth, 0, 0, 0);
        return ans;
    }
};

File: JAVA/1105. Filling Bookcase Shelves.java
class Solution {
    int Width;
    int[][] a = new int[1001][1001];
    public int minHeightShelves(int[][] books, int shelfWidth) {
        int remW = shelfWidth;
        for (int[] row : a) 
            Arrays.fill(row, -1); 
        Width = shelfWidth;
        return solve(books, 0, remW,0);
    }
    public int solve(int[][] books,int i,int remW,int maxH)
    {
        if(i>=books.length)
        return maxH;
        if(a[i][remW]!=-1)
        return a[i][remW];
        int bookH = books[i][1];
        int bookW = books[i][0];
        int skip = Integer.MAX_VALUE;
        int keep = Integer.MAX_VALUE;
        if(bookW<=remW)
        {
           keep = solve(books,i+1,remW-bookW,Math.max(maxH,bookH));
        }
         skip  = maxH + solve(books,i+1,Width-bookW,bookH);
        return a[i][remW] = Math.min(keep,skip);
    }
}
